diff --git a/haskell/private/haskell_impl.bzl b/haskell/private/haskell_impl.bzl
index 3bc7cd8..865dae6 100644
--- a/haskell/private/haskell_impl.bzl
+++ b/haskell/private/haskell_impl.bzl
@@ -146,8 +146,15 @@ def _expand_make_variables(name, ctx, strings):
         ctx.attr.plugins,
         ctx.attr.tools,
     ]
-    for attr in label_attrs:
-        strings = [ctx.expand_location(str, attr) for str in strings]
+    # Deduplicate targets. Targets could be duplicated between attributes, e.g.
+    # srcs and extra_srcs. ctx.expand_location fails if any target occurs
+    # multiple times.
+    targets = {
+        target.label: target
+        for attr in label_attrs
+        for target in attr
+    }.values()
+    strings = [ctx.expand_location(str, targets = targets) for str in strings]
     strings = [ctx.expand_make_variables(name, str, {}) for str in strings]
     return strings
 
